File: agent/services/ops_models.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any, Literal

OpsErrorCode = Literal[
    "git_not_found",
    "git_not_repository",
    "git_timeout",
    "git_command_failed",
    "git_detached_head",
    "git_dirty_worktree",
    "git_conflict",
    "git_operation_in_progress",
    "git_no_upstream",
    "git_remote_not_allowed",
    "git_branch_not_allowed",
    "git_path_state_invalid",
    "git_untracked_discard_denied",
    "git_nothing_to_commit",
    "invalid_commit_message",
    "docker_not_found",
    "docker_unreachable",
    "docker_permission_denied",
    "docker_boundary_not_configured",
    "docker_container_not_registered",
    "compose_plugin_missing",
    "compose_file_invalid",
    "compose_project_not_registered",
    "workspace_not_allowed",
    "path_not_allowed",
    "approval_required",
    "policy_denied",
    "output_truncated",
]


READ_ACTIONS = frozenset(
    {
        "status",
        "diff",
        "changes",
        "history",
        "branches",
        "remotes",
        "activity",
        "list",
        "logs",
        "inspect_light",
        "stats",
        "info",
        "config",
    }
)
MUTATING_ACTIONS = frozenset(
    {
        "stage",
        "unstage",
        "discard",
        "commit",
        "fetch",
        "pull_ff_only",
        "push",
        "start",
        "stop",
        "restart",
        "pull",
        "up",
        "down",
    }
)
DANGEROUS_ACTIONS = frozenset(
    {"force_push", "reset_hard", "clean", "rm_volume", "prune", "compose_down_volumes", "docker_socket_mount_enable"}
)


@dataclass(frozen=True)
class SerializableDto:
    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True)
class OpsError(SerializableDto):
    code: str
    message: str
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class OpsActionResult(SerializableDto):
    ok: bool
    action: str
    target_id: str = ""
    decision: str = "allow"
    error: OpsError | None = None
    approval_id: str | None = None
    audit_ref: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class GitChangedFile(SerializableDto):
    path: str
    original_path: str = ""
    index_status: str = ""
    worktree_status: str = ""
    staged: bool = False
    unstaged: bool = False
    untracked: bool = False
    conflicted: bool = False
    renamed: bool = False
    deleted: bool = False
    binary: bool = False
    additions: int = 0
    deletions: int = 0


@dataclass(frozen=True)
class GitCommitSummary(SerializableDto):
    sha: str
    subject: str
    short_sha: str = ""
    author_name: str = ""
    author_email: str = ""
    authored_at: str = ""
    parents: list[str] = field(default_factory=list)
    refs: list[str] = field(default_factory=list)


@dataclass(frozen=True)
class GitStatus(SerializableDto):
    workspace_id: str
    branch: str = ""
    head_sha: str = ""
    upstream: str = ""
    remote_name: str = ""
    detached: bool = False
    ahead: int = 0
    behind: int = 0
    operation_state: str = "idle"
    dirty: bool = False
    conflict_count: int = 0
    staged_count: int = 0
    unstaged_count: int = 0
    untracked_count: int = 0
    can_commit: bool = False
    can_pull: bool = False
    can_push: bool = False
    truncated: bool = False
    changed_files: list[GitChangedFile] = field(default_factory=list)
    recent_commits: list[GitCommitSummary] = field(default_factory=list)
    error: OpsError | None = None


@dataclass(frozen=True)
class GitDiffStat(SerializableDto):
    path: str
    additions: int = 0
    deletions: int = 0
    binary: bool = False


@dataclass(frozen=True)
class GitDiff(SerializableDto):
    workspace_id: str
    cached: bool = False
    path: str = ""
    scope: str = "unstaged"
    head_sha: str = ""
    diff: str = ""
    staged_diff: str = ""
    unstaged_diff: str = ""
    untracked_diff: str = ""
    stats: list[GitDiffStat] = field(default_factory=list)
    additions: int = 0
    deletions: int = 0
    files_changed: int = 0
    truncated: bool = False
    error: OpsError | None = None


@dataclass(frozen=True)
class GitChanges(SerializableDto):
    workspace_id: str
    items: list[GitChangedFile] = field(default_factory=list)
    count: int = 0
    staged_count: int = 0
    unstaged_count: int = 0
    untracked_count: int = 0
    conflict_count: int = 0
    truncated: bool = False
    error: OpsError | None = None


@dataclass(frozen=True)
class GitHistory(SerializableDto):
    workspace_id: str
    items: list[GitCommitSummary] = field(default_factory=list)
    count: int = 0
    limit: int = 50
    offset: int = 0
    has_more: bool = False
    error: OpsError | None = None


@dataclass(frozen=True)
class GitBranch(SerializableDto):
    name: str
    current: bool = False
    remote: bool = False
    upstream: str = ""
    ahead: int = 0
    behind: int = 0
    sha: str = ""
    last_commit_sha: str = ""
    last_commit_subject: str = ""
    last_commit_at: str = ""


@dataclass(frozen=True)
class GitBranches(SerializableDto):
    workspace_id: str
    items: list[GitBranch] = field(default_factory=list)
    count: int = 0
    error: OpsError | None = None


@dataclass(frozen=True)
class GitRemote(SerializableDto):
    name: str
    fetch_url: str = ""
    push_url: str = ""


@dataclass(frozen=True)
class GitRemotes(SerializableDto):
    workspace_id: str
    items: list[GitRemote] = field(default_factory=list)
    count: int = 0
    error: OpsError | None = None


@dataclass(frozen=True)
class GitActivityEvent(SerializableDto):
    id: str
    timestamp: str
    actor: str
    action: str
    operation: str = ""
    outcome: str = "unknown"
    source: str = "ops"
    workspace_id: str = ""
    task_id: str = ""
    goal_id: str = ""
    trace_id: str = ""
    approval_id: str = ""
    summary: str = ""


@dataclass(frozen=True)
class GitActivity(SerializableDto):
    workspace_id: str
    items: list[GitActivityEvent] = field(default_factory=list)
    count: int = 0
    error: OpsError | None = None


@dataclass(frozen=True)
class DockerEngineStatus(SerializableDto):
    available: bool
    boundary: str = "disabled"
    docker_version: str = ""
    compose_available: bool = False
    platform_hint: str = ""
    engine: dict[str, Any] = field(default_factory=dict)
    error: OpsError | None = None


@dataclass(frozen=True)
class DockerContainerSummary(SerializableDto):
    id: str
    name: str
    image: str
    status: str
    health: str = ""
    ports: str = ""
    labels: dict[str, str] = field(default_factory=dict)
    compose_project: str = ""
    uptime: str = ""
    state: str = ""
    command: str = ""
    created_at: str = ""
    size: str = ""
    networks: list[str] = field(default_factory=list)
    mounts: list[str] = field(default_factory=list)
    registered: bool = True
    managed: bool = False
    allowed_actions: list[str] = field(default_factory=list)


@dataclass(frozen=True)
class ComposeServiceStatus(SerializableDto):
    name: str
    state: str = ""
    health: str = ""
    exit_code: str = ""
    ports: str = ""
    container_id: str = ""
    image: str = ""
    command: str = ""
    created_at: str = ""


@dataclass(frozen=True)
class ComposeProjectSummary(SerializableDto):
    project_id: str
    name: str
    project_directory: str
    compose_files: list[str]
    profiles: list[str] = field(default_factory=list)
    available_profiles: list[str] = field(default_factory=list)
    marker: str = "preferred"
    category: str = "dev"
    allowed_actions: list[str] = field(default_factory=list)
    services: list[ComposeServiceStatus] = field(default_factory=list)
    error: OpsError | None = None

File: frontend-angular/src/app/features/operations/ops.models.ts
export interface OpsError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export interface OpsActionResult {
  ok: boolean;
  action: string;
  target_id?: string;
  decision?: 'allow' | 'approval_required' | 'policy_denied' | string;
  approval_id?: string | null;
  audit_ref?: string | null;
  metadata?: Record<string, unknown>;
  error?: OpsError | null;
}

export interface GitWorkspaceSummary {
  workspace_id: string;
  label?: string;
  root?: string;
  is_default?: boolean;
  repository?: boolean;
  source?: string;
}

export interface GitChangedFile {
  path: string;
  original_path?: string;
  index_status?: string;
  worktree_status?: string;
  staged?: boolean;
  unstaged?: boolean;
  untracked?: boolean;
  conflicted?: boolean;
  renamed?: boolean;
  deleted?: boolean;
  binary?: boolean;
  additions?: number;
  deletions?: number;
}

export interface GitCommitSummary {
  sha: string;
  short_sha?: string;
  subject: string;
  author_name?: string;
  author_email?: string;
  authored_at?: string;
  parents?: string[];
  refs?: string[];
}

export interface GitBranchSummary {
  name: string;
  current?: boolean;
  remote?: boolean;
  upstream?: string;
  ahead?: number;
  behind?: number;
  sha?: string;
  last_commit_sha?: string;
  last_commit_subject?: string;
  last_commit_at?: string;
}

export interface GitRemoteSummary {
  name: string;
  fetch_url?: string;
  push_url?: string;
}

export interface GitActivityEntry {
  id?: string;
  timestamp?: string;
  actor?: string;
  operation?: string;
  action?: string;
  outcome?: string;
  summary?: string;
  source?: string;
  task_id?: string | null;
  goal_id?: string | null;
  audit_ref?: string | null;
  trace_id?: string | null;
  approval_id?: string | null;
  metadata?: Record<string, unknown>;
}

export interface GitStatus {
  workspace_id: string;
  branch: string;
  head_sha?: string;
  upstream: string;
  remote_name: string;
  ahead?: number;
  behind?: number;
  dirty: boolean;
  detached?: boolean;
  operation_state?: string;
  conflict_count?: number;
  staged_count?: number;
  unstaged_count?: number;
  untracked_count?: number;
  can_commit?: boolean;
  can_pull?: boolean;
  can_push?: boolean;
  changed_files: GitChangedFile[];
  recent_commits: GitCommitSummary[];
  branches?: GitBranchSummary[];
  remotes?: GitRemoteSummary[];
  error?: OpsError | null;
}

export interface GitDiff {
  workspace_id: string;
  cached: boolean;
  path: string;
  diff: string;
  scope?: 'unstaged' | 'staged' | 'combined' | string;
  staged_diff?: string;
  unstaged_diff?: string;
  untracked_diff?: string;
  truncated: boolean;
  additions?: number;
  deletions?: number;
  files_changed?: number;
  error?: OpsError | null;
}

export interface DockerEngineStatus {
  available: boolean;
  boundary: string;
  docker_version: string;
  api_version?: string;
  os?: string;
  architecture?: string;
  compose_available: boolean;
  platform_hint: string;
  error?: OpsError | null;
}

export interface DockerContainerSummary {
  id: string;
  name: string;
  image: string;
  image_id?: string;
  status: string;
  state?: string;
  health?: string;
  ports?: string;
  labels?: Record<string, string>;
  compose_project?: string;
  compose_service?: string;
  created_at?: string;
  command?: string;
  size?: string;
  networks?: Array<string | { name?: string; ip_address?: string; gateway?: string }>;
  mounts?: Array<string | { source?: string; destination?: string; type?: string; read_only?: boolean }>;
  registered?: boolean;
  allowed_actions?: string[];
  uptime?: string;
}

export interface DockerContainerDetails {
  ok?: boolean;
  id?: string;
  name?: string;
  image?: string;
  status?: string;
  command?: string;
  inspect?: {
    state?: Record<string, unknown>;
    health?: Record<string, unknown>;
    resources?: Record<string, unknown>;
    mounts?: Array<Record<string, unknown>>;
    networks?: Record<string, unknown> | Array<Record<string, unknown>>;
    ports?: Record<string, unknown>;
    restart_policy?: Record<string, unknown> | string;
    [key: string]: unknown;
  };
  restart_policy?: string;
  mounts?: Array<{ source?: string; destination?: string; type?: string; read_only?: boolean }>;
  networks?: Array<{ name?: string; ip_address?: string; gateway?: string }>;
  environment_keys?: string[];
}

export interface DockerContainerStats {
  id?: string;
  name?: string;
  cpu_percent?: number | string;
  memory_usage?: number | string;
  memory_limit?: number | string;
  memory_percent?: number | string;
  network_rx?: number;
  network_tx?: number;
  net_io?: string | Record<string, unknown>;
  network_io?: string | Record<string, unknown>;
  block_io?: string | Record<string, unknown>;
  block_read?: number;
  block_write?: number;
  pids?: number | string;
  error?: OpsError | null;
}

export interface DockerInfo {
  id?: string;
  name?: string;
  server_version?: string;
  operating_system?: string;
  architecture?: string;
  kernel_version?: string;
  storage_driver?: string;
  driver?: string;
  cgroup_driver?: string;
  cpus?: number;
  memory_total?: number;
  memory_bytes?: number;
  containers?: number;
  containers_running?: number;
  containers_paused?: number;
  containers_stopped?: number;
  images?: number;
  error?: OpsError | null;
}

export interface DockerImageSummary {
  id: string;
  repository?: string;
  tag?: string;
  digest?: string;
  created_at?: string;
  size?: number | string;
  containers?: number;
}

export interface DockerNetworkSummary {
  id: string;
  name: string;
  driver?: string;
  scope?: string;
  internal?: boolean | string;
  attachable?: boolean | string;
  ipv6?: boolean | string;
  containers?: number | string;
}

export interface DockerVolumeSummary {
  name: string;
  driver?: string;
  scope?: string;
  mountpoint?: string;
  labels?: Record<string, string>;
  in_use?: boolean;
  links?: string[];
  size?: number | string;
}

export interface DockerDiskUsage {
  images?: { count?: number | string; active?: number | string; size?: number | string; reclaimable?: number | string };
  containers?: { count?: number | string; active?: number | string; size?: number | string; reclaimable?: number | string };
  volumes?: { count?: number | string; active?: number | string; size?: number | string; reclaimable?: number | string };
  build_cache?: { count?: number | string; active?: number | string; size?: number | string; reclaimable?: number | string };
  error?: OpsError | null;
}

export interface ComposeServiceStatus {
  name: string;
  container_id?: string;
  state: string;
  status?: string;
  health?: string;
  exit_code?: string;
  ports?: string;
  image?: string;
}

export interface ComposeProjectSummary {
  project_id: string;
  name: string;
  project_directory: string;
  compose_files: string[];
  profiles: string[];
  available_profiles?: string[];
  marker: string;
  category: string;
  allowed_actions?: string[];
  services: ComposeServiceStatus[];
  error?: OpsError | null;
}

export interface OpsTextResult {
  ok: boolean;
  logs?: string;
  config?: string;
  stderr?: string;
  truncated?: boolean;
  error?: OpsError | null;
}

export interface OpsApprovalRequest {
  request_id: string;
  task_id?: string | null;
  goal_id?: string | null;
  trace_id?: string | null;
  tool_name: string;
  digest_prefix: string;
  target_fingerprint_prefix?: string;
  risk_class?: string;
  k_class?: string | null;
  governance_mode?: string;
  status: 'pending' | 'granted' | 'denied' | 'expired' | 'consumed' | 'superseded' | string;
  scope_summary?: Record<string, unknown>;
  created_at?: number;
  expires_at?: number | null;
  decided_at?: number | null;
  decided_by?: string | null;
  decision_reason?: string | null;
  consumed_at?: number | null;
  has_content_payload?: boolean;
}

export type OpsArea = 'git' | 'docker' | 'compose' | 'approvals';
